Subcommands, custom tags, modified discordboosters

RyanLandDev opened the pull request:
- Created new directories for impl commands to be more organized, no difference front-end.
- Redid DiscordBoostersCommand to add a small description and use mentions instead of names (doesn't actually ping the user), aswell as include a formatted time for how long this user has been boosting the server for.
- Added a new Tags system, which allows Experts+ to add and modify custom command tags, which anyone can then use.

Owen1212055 commented on src/main/java/com/diamondfire/helpbot/bot/command/impl/stats/top/DiscordBoostersCommand.java:
```diff
@@ -48,22 +49,34 @@ public void run(CommandEvent event) {
-}
-
-
+    private String getFormattedBoosters(List<Member> members) {
+        if (members.size() == 0) return "*None*";
+        else {
+            SimpleDateFormat format = new SimpleDateFormat("y'y'M'm'D'd'");
+            String result = "";
+            
+            for (Member member : members) {
+                String timeBoosted = format.format(member.getTimeBoosted());
+                result += "<@" + member.getId() + "> - " + timeBoosted + "\n";
+            }
+            
+            return result;
+        }
+    }
```
This is a more of a readability opinion, but instead of having a result and appending newline do something like this.
```java
List<String> elements = new ArrayList<>();
for (Member member : members) {
    String timeBoosted = ....
    elements.add(member.getAsMention() + " - " + timeBoosted);
}

return Strings.join("\n", elements);
```



```

Owen1212055 replied:
And make this method private static

Owen1212055 commented on src/main/java/com/diamondfire/helpbot/bot/command/impl/other/tag/Tag.java:
```diff
@@ -0,0 +1,98 @@
+        channel.sendMessage(embed.build()).queue();
+    }
+    
+    @Override
+    public Tag clone() throws CloneNotSupportedException {
+        return (Tag) super.clone();
+    }
```
Don't use cloning

Owen1212055 commented on src/main/java/com/diamondfire/helpbot/bot/command/impl/other/tag/Tag.java:
```diff
@@ -0,0 +1,98 @@
+    }
+    
+    public void set(TagProperty property, String newValue) {
+        if (property == TagProperty.ACTIVATOR) setActivator(newValue);
+        else if (property == TagProperty.TITLE) setTitle(newValue);
+        else if (property == TagProperty.RESPONSE) setResponse(newValue);
+        else if (property == TagProperty.IMAGE) setImage(newValue);
```
Use an abstract method in the enum. Where it's like TagProperty.IMAGE.edit(newValue);

Owen1212055 commented:
A general comment, but all the tag logic should go in it's own package in the ``sys`` package. So this would include the enum, tag itself, and exceptions.

Owen1212055 commented:
Additionally, should the tag commands be separate or should it be considered combining them?

Like ``tag add/remove/edit``?

RyanLandDev commented:
all fixed (unpushed), except a better TagHandler / Tag cache + a proper subcommands implementation

RyanLandDev commented:
- All requested changes complete
- Better Tag cache
- Added subcommands system
- Adapted the tag commands to use the new subcommands
- Removed the 'tag' alias from the ?tags command in favor of ^

RyanLandDev commented:
Subcommands have been reworked
